Vimgoto additions

AndrewRadev opened the pull request:
This PR adds two additional `gf` targets for Vimscript files. I only recently noticed that there was a built-in custom gf implementation for these because it overrode my own custom include expression 😅. These two features are taken from that one.

- Pressing `gf` on `foo#bar#Baz()` will look for `autoload/foo/bar.vim` in the runtimepath
- Pressing `gf` on `runtime some/path.vim` will look for the path in the runtimepath

I've tried to follow a similar code style to the existing code. A few things look duplicated to me, which is why I have a separate branch that extracts some helper functions and makes the code a bit simpler, I think: https://github.com/AndrewRadev/vim/compare/vimgoto-additions...AndrewRadev:vim:vimgoto-additions-refactoring. I can merge that one as well into this PR if the changes seem reasonable. (Update: merged)

AndrewRadev committed:
Implement jumping to autoloaded functions

Signed-off-by: Andrew Radev <andrey.radev@gmail.com>

AndrewRadev committed:
Implement jumping to runtime-d files

Signed-off-by: Andrew Radev <andrey.radev@gmail.com>

AndrewRadev committed:
Update header comment

Signed-off-by: Andrew Radev <andrey.radev@gmail.com>

AndrewRadev committed:
Extract editcmd processing to the `Open` function

Signed-off-by: Andrew Radev <andrey.radev@gmail.com>

AndrewRadev committed:
Extract some helper functions

Signed-off-by: Andrew Radev <andrey.radev@gmail.com>

AndrewRadev commented on runtime/autoload/vimgoto.vim:
```diff
@@ -150,8 +204,16 @@ def Open(what: any, how: string) #{{{2
+    if search_pattern != ''
+        var escaped_pattern = escape(search_pattern, '\#'' ')
+        cmd = $'+silent\ call\ search(''{escaped_pattern}'')'
+    endif
```
This code adds an additional command to search for the `foo#bar` pattern for functions. That way, pressing `gf` jumps to the right function, and not just to the file. It uses a `call search(` because the built-in `+/<pattern>` modifier only jumps to a line, not to the specific pattern, so a `fun\%[ction]!=\zs{curfunc}` will put the cursor on the beginning on the line instead of after the `\zs` anchor. If this seems too complicated or error-prone, I can switch to `+/` here.

AndrewRadev commented on runtime/autoload/vimgoto.vim:
```diff
@@ -150,8 +204,16 @@ def Open(what: any, how: string) #{{{2
+
+    if search_pattern == ''
+        cursor(1, 1)
+    endif
```
Personally, I would avoid skipping to line 1, column 1, I feel like it's better to leave it to the user. Most people are probably going to use the default "jump to the last known position in the file" autocommand, and this disrupts that functionality. But maybe it was intended to fix some issue, I don't know.

habamax replied:
`cursor(1, 1)` looks redundant here, indeed. It should jump to it by default.

AndrewRadev replied:
Removed 👍

AndrewRadev committed:
Remove unnecessary cursor(1, 1)

It might be a work-around to avoid a "jump to the last known location"
autocommand, but that shouldn't be necessary.

Signed-off-by: Andrew Radev <andrey.radev@gmail.com>

dkearns commented on runtime/autoload/vimgoto.vim:
```diff
@@ -20,6 +20,11 @@ export def Find(editcmd: string) #{{{2
         return
     endif
 
+    if curline =~ '^\s*\%(:\s*\)\=runtime!\=\s'
```
```suggestion
    if curline =~ '^\s*\%(:\s*\)\=ru\%[ntime]!\=\s'
```

AndrewRadev committed:
Improve `runtime` regex

Co-authored-by: dkearns <dougkearns@gmail.com>

dkearns commented on runtime/autoload/vimgoto.vim:
```diff
@@ -45,22 +64,34 @@ def HandlePackaddLine(editcmd: string, curline: string) #{{{2
+enddef
+
+def HandleRuntimeLine(editcmd: string, curline: string) #{{{2
+    var pat: string = '\s*\%(:\s*\)\=runtime!\=\s\+\zs\S\+\>\ze'
```
Sorry I forgot to add a note about this one too.

```suggestion
    var pat: string = '\s*\%(:\s*\)\=ru\%[ntime]!\=\s\+\zs\S\+\>\ze'
```

Whitespace after the bang is not actually required nor after the command provided the filename isn't an alpha character.   I've noticed some examples of these in the wild but I'm not sure they're worth supporting.

`:runtime` supports multiple file arguments as well, if you want to tackle that at some point.

AndrewRadev replied:
Hm, I hadn't realized these tricks with `runtime!`. There's also the option of a second argument between the `runtime!` and the filename for `START`, `OPT`, etc. I've pushed some changes that should handle this, and multiple files. I'm taking `<cfile>` as an additional argument and checking whether we should follow that one. I think for `runtime` in particular, you can't really have unescaped spaces in those files, so `<cfile>` should manage?

dkearns commented on runtime/autoload/vimgoto.vim:
```diff
@@ -30,6 +35,20 @@ export def Find(editcmd: string) #{{{2
+        var resolved_path = globpath(&runtimepath, path)
+
+        if resolved_path != ''
+            var function_pattern: string = $'^\s*fun\%[ction]!\=\s\+\zs{curfunc}('
```
This could also have a command colon but it's obviously rare.

AndrewRadev replied:
Yeah, you're right, added the same prefix as the other commands.

AndrewRadev committed:
Improve second `runtime` regex

Co-authored-by: dkearns <dougkearns@gmail.com>

AndrewRadev committed:
Handle more cases for runtime!

Signed-off-by: Andrew Radev <andrey.radev@gmail.com>

AndrewRadev committed:
Handle case with no space after runtime!

Signed-off-by: Andrew Radev <andrey.radev@gmail.com>

AndrewRadev committed:
Handle function definitions with a starting `:`
